Add bn254 native contract bindings and tests

Jim8y opened the pull request:
# Description

Proposal: https://github.com/neo-project/proposals/pull/205.

Adds BN254 native CryptoLib support using a managed implementation and the same opaque point interface shape as the existing BLS12-381 methods.

## Why BN254 matters

BN254 is one of the most widely deployed pairing-friendly curves in the zero-knowledge ecosystem. Groth16 and many ZK systems and zkVM proof pipelines use BN254 directly or produce BN254-based proofs for efficient on-chain verification. Ethereum exposes BN254 precompiles for the same reason, so supporting compatible encodings also improves interoperability with existing proving stacks and tooling.

Native CryptoLib support lets Neo smart contracts verify these proofs without implementing large-field arithmetic and pairing operations in contract code. This enables practical privacy proofs, rollup and bridge verification, verifiable computation, identity and credential proofs, and zkVM result verification with deterministic consensus behavior and predictable fees.

The exposed native methods are:

- `bn254Serialize(InteropInterface point) -> ByteArray`
- `bn254Deserialize(ByteArray data) -> InteropInterface`
- `bn254Add(InteropInterface x, InteropInterface y) -> InteropInterface`
- `bn254Equal(InteropInterface x, InteropInterface y) -> Boolean`
- `bn254Mul(InteropInterface point, ByteArray scalar) -> InteropInterface`
- `bn254Pairing(InteropInterface g1, InteropInterface g2) -> InteropInterface`

The implementation uses Ethereum-compatible BN254 byte layouts for serialized G1, G2, scalar, and GT values, validates canonical coordinates and subgroup membership, and returns typed interop values so contracts can deserialize points once, perform multiple operations, and serialize only at the boundary.

The native manifest snapshot is refreshed so CryptoLib exposes the new entry points after the configured hardfork.

Fixes # (issue)

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] `dotnet test tests/Neo.UnitTests/Neo.UnitTests.csproj --filter "FullyQualifiedName~BN254|FullyQualifiedName~Bn254" --verbosity minimal`

Test Configuration:

- OS: macOS arm64
- .NET SDK: 10.0.108
- Tests include typed CryptoLib operations, serialization round trips, subgroup checks, non-canonical coordinate rejection, and Ethereum precompile-compatible vectors.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

Wi1l-B0t commented on src/Neo/Cryptography/BN254.cs:
```diff
@@ -0,0 +1,277 @@
+        public const int PairInputLength = 192;
+
+        private static readonly object s_sync = new();
+        private static bool s_initialized;
```
This is not thread safe. The `s_initialized` must be volatile.

```suggestion
        private static volatile bool s_initialized;
```

AnnaShaleva reviewed (commented):
Need some more time to verify the math.

Wi1l-B0t commented:
No progress?

shargon commented on src/Neo/SmartContract/Native/CryptoLib.BN254.cs:
```diff
@@ -0,0 +1,49 @@
+        {
+            ArgumentNullException.ThrowIfNull(input);
+            if (input.Length != BN254.G1EncodedLength * 2)
+                throw new ArgumentException("Invalid BN254 add input length", nameof(input));
```
Move this validation to `Add` method?

shargon commented:
> I briefly read through the implementation, and it looks compatible with the Solidity/EVM precompile interface.
> 
> To verify the equivalence, please consider using the EVM test data, ref https://github.com/ethereum/go-ethereum/tree/v1.16.7/core/vm/testdata/precompiles.

@Jim8y  Nice, could you add them to test?

txhsl reviewed (commented):
Only a performance concern, otherwise all good to me.

AnnaShaleva reviewed (commented):
Please, consider also to update the price: https://github.com/neo-project/neo/pull/4185/files#r2399242679.

AnnaShaleva commented on src/Neo/SmartContract/Native/CryptoLib.BN254.cs:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        [ContractMethod(Hardfork.HF_Faun, CpuFee = 1 << 19, Name = "bn254_add")]
+        public static byte[] Bn254AddRaw(byte[] input)
```
Please, remove `Bn254*Raw` methods. Standard methods that work with InteropInterface are sufficient.

Jim8y committed:
feat(crypto): add BN254 (alt_bn128) curve support to CryptoLib

Add the BN254 (alt_bn128) pairing-friendly curve to the CryptoLib
native contract as a fully managed implementation built on
System.Numerics.BigInteger, with no native or unsafe dependencies.

The curve is exposed through six hardfork-gated methods that mirror the
existing BLS12-381 surface and return point/GT handles via
InteropInterface: bn254Serialize, bn254Deserialize, bn254Add, bn254Mul,
bn254Pairing and bn254Equal.

- G1/G2 use the EIP-196/197 big-endian encodings; GT is a 384-byte
  canonical F_p12 element.
- G2 inputs are validated for on-curve and subgroup membership, and all
  field elements must be canonical (< p), matching the EVM precompiles
  and SP1 verifier.
- The pairing is priced per input pair to bound execution cost.
- All methods are activated under the HF_Gorgon hardfork.

Includes unit tests and EIP-197 test vectors for point addition, scalar
multiplication and pairing, plus GT serialization round-trips and
equality checks.

Jim8y committed:
Merge master-n3 into feature/bn254-curve-support

Jim8y committed:
Merge remote-tracking branch 'origin/master-n3' into feature/bn254-curve-support

Jim8y committed:
Validate BN254 scalar length
